sklearn/mixture/_gaussian_mixture_ic.py

```diff
@@ -0,0 +1,776 @@
+            affinity used in Agglomerative Clustering
+        linkage : {'ward','complete','average','single'}
+            linkage used in Agglomerative Clustering
+        covariance_type : {'full', 'tied', 'diag', 'spherical'}
+            covariance type used in GMM
+        reg_covar : float
+            regularization used in GMM
+
+    n_features_in_ : int
+        Number of features seen during :term:`fit`.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.mixture import GaussianMixtureIC
+    >>> X = np.array([[1, 2], [1, 4], [1, 0], [10, 2], [10, 4], [10, 0]])
+    >>> gmIC = GaussianMixtureIC(max_components=4, random_state=0)
+    >>> gmIC.fit_predict(X)
+    array([0, 0, 0, 1, 1, 1])
+    >>> print(gmIC.n_components_)
+    2
+
+    See Also
+    --------
+    GaussianMixture : fit Gaussian mixture model
+    BayesianGaussianMixture : Gaussian mixture model fit with a variational
+    inference
+
+    Notes
+    -----
+    This algorithm was strongly inspired by mclust, a clustering package in R
+
+    References
+    ----------
+    .. [1] Jeffrey D. Banfield and Adrian E. Raftery. Model-based gaussian and
+       non-gaussian clustering. Biometrics, 49:803–821, 1993.
+
+    .. [2] Abhijit Dasgupta and Adrian E. Raftery. Detecting features in
+       spatial point processes with clutter via model-based clustering.
+       Journal of the American Statistical Association, 93(441):294–302, 1998.
+    """
+
+    def __init__(
+        self,
+        min_components=2,
+        max_components=10,
+        affinity="all",
+        linkage="all",
+        covariance_type="all",
+        random_state=None,
+        label_init=None,
+        max_iter=100,
+        verbose=0,
+        criterion="bic",
+        max_agglom_size=2000,
+        n_jobs=None,
+    ):
+
+        self.min_components = min_components
+        self.max_components = max_components
+        self.affinity = affinity
+        self.linkage = linkage
+        self.covariance_type = covariance_type
+        self.random_state = random_state
+        self.max_iter = max_iter
+        self.label_init = label_init
+        self.verbose = verbose
+        self.criterion = criterion
+        self.max_agglom_size = max_agglom_size
+        self.n_jobs = n_jobs
+
+    def _check_multi_comp_inputs(self, input, name, default):
+        if isinstance(input, (np.ndarray, list)):
+            input = list(np.unique(input))
+        elif isinstance(input, str):
+            if input not in default:
+                msg = f"{name} is {input} but must be one of {default}."
+                raise ValueError(msg)
+            if input != "all":
+                input = [input]
+        else:
+            msg = (
+                f"{name} is a {type(input)} but must be a numpy array, "
+                "a list, or a string."
+            )
+            raise TypeError(msg)
+        return input
+
+    def _check_parameters(self):
```

Comment:
@tliu68 [randomness] I'd add a [`check_random_state`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_random_state.html#sklearn.utils.check_random_state) see [here](https://github.com/scikit-learn/scikit-learn/blob/95119c13af77c76e150b753485c662b7c52a41a2/sklearn/cluster/_kmeans.py#L985) for example